File: backend/internal/handlers/restaurant_info.go
package handlers

import (
	"database/sql"
	"net/http"
	"regexp"
	"strings"
	"time"

	"pos-public/internal/models"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

// RestaurantInfoHandler handles admin operations for restaurant information
type RestaurantInfoHandler struct {
	db *sql.DB
}

// NewRestaurantInfoHandler creates a new restaurant info handler
func NewRestaurantInfoHandler(db *sql.DB) *RestaurantInfoHandler {
	return &RestaurantInfoHandler{db: db}
}

// UpdateRestaurantInfoRequest represents the request body for updating restaurant info
type UpdateRestaurantInfoRequest struct {
	Name          string   `json:"name" binding:"required"`
	Tagline       *string  `json:"tagline"`
	Description   *string  `json:"description"`
	Address       *string  `json:"address"`
	City          *string  `json:"city"`
	PostalCode    *string  `json:"postal_code"`
	Country       *string  `json:"country"`
	Phone         *string  `json:"phone"`
	Email         *string  `json:"email"`
	WhatsApp      *string  `json:"whatsapp"`
	MapLatitude   *float64 `json:"map_latitude"`
	MapLongitude  *float64 `json:"map_longitude"`
	GoogleMapsURL *string  `json:"google_maps_url"`
	InstagramURL  *string  `json:"instagram_url"`
	FacebookURL   *string  `json:"facebook_url"`
	TwitterURL    *string  `json:"twitter_url"`
	LogoURL       *string  `json:"logo_url"`
	HeroImageURL  *string  `json:"hero_image_url"`
}

// UpdateOperatingHoursRequest represents the request for updating operating hours
type UpdateOperatingHoursRequest struct {
	Hours []OperatingHourUpdate `json:"hours" binding:"required,dive"`
}

type OperatingHourUpdate struct {
	DayOfWeek int    `json:"day_of_week" binding:"min=0,max=6"`
	OpenTime  string `json:"open_time"`
	CloseTime string `json:"close_time"`
	IsClosed  bool   `json:"is_closed"`
}

// UpdateRestaurantInfo updates the restaurant information
// Endpoint: PUT /api/v1/admin/restaurant-info
func (h *RestaurantInfoHandler) UpdateRestaurantInfo(c *gin.Context) {
	var req UpdateRestaurantInfoRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Message: "Invalid request format",
			Error:   stringPtr(err.Error()),
		})
		return
	}

	// Validate required fields
	req.Name = strings.TrimSpace(req.Name)
	if req.Name == "" {
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Message: "Restaurant name is required",
			Error:   stringPtr("name_required"),
		})
		return
	}

	// Validate email format if provided
	if req.Email != nil && *req.Email != "" {
		email := strings.TrimSpace(*req.Email)
		req.Email = &email
		if !isValidEmail(*req.Email) {
			c.JSON(http.StatusBadRequest, models.APIResponse{
				Success: false,
				Message: "Invalid email format",
				Error:   stringPtr("invalid_email"),
			})
			return
		}
	}

	// Validate phone format if provided
	if req.Phone != nil && *req.Phone != "" {
		phone := strings.TrimSpace(*req.Phone)
		req.Phone = &phone
		if !isValidPhone(*req.Phone) {
			c.JSON(http.StatusBadRequest, models.APIResponse{
				Success: false,
				Message: "Invalid phone format",
				Error:   stringPtr("invalid_phone"),
			})
			return
		}
	}

	// Get restaurant info ID (singleton table)
	var infoID uuid.UUID
	err := h.db.QueryRow("SELECT id FROM restaurant_info LIMIT 1").Scan(&infoID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Message: "Failed to retrieve restaurant info",
			Error:   stringPtr(err.Error()),
		})
		return
	}

	// Update restaurant info
	updateQuery := `
		UPDATE restaurant_info SET
			name = $1, tagline = $2, description = $3, address = $4, city = $5,
			postal_code = $6, country = $7, phone = $8, email = $9, whatsapp = $10,
			map_latitude = $11, map_longitude = $12, google_maps_url = $13,
			instagram_url = $14, facebook_url = $15, twitter_url = $16,
			logo_url = $17, hero_image_url = $18, updated_at = NOW()
		WHERE id = $19
	`

	_, err = h.db.Exec(
		updateQuery,
		req.Name,
		req.Tagline,
		req.Description,
		req.Address,
		req.City,
		req.PostalCode,
		req.Country,
		req.Phone,
		req.Email,
		req.WhatsApp,
		req.MapLatitude,
		req.MapLongitude,
		req.GoogleMapsURL,
		req.InstagramURL,
		req.FacebookURL,
		req.TwitterURL,
		req.LogoURL,
		req.HeroImageURL,
		infoID,
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Message: "Failed to update restaurant information",
			Error:   stringPtr(err.Error()),
		})
		return
	}

	c.JSON(http.StatusOK, models.APIResponse{
		Success: true,
		Message: "Restaurant information updated successfully",
		Data: gin.H{
			"id": infoID,
		},
	})
}

// UpdateOperatingHours updates the restaurant operating hours
// Endpoint: PUT /api/v1/admin/operating-hours
func (h *RestaurantInfoHandler) UpdateOperatingHours(c *gin.Context) {
	var req UpdateOperatingHoursRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Message: "Invalid request format",
			Error:   stringPtr(err.Error()),
		})
		return
	}

	// Validate hours count (must be 7 days)
	if len(req.Hours) != 7 {
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Message: "Operating hours must include all 7 days of the week",
			Error:   stringPtr("invalid_hours_count"),
		})
		return
	}

	// Validate time formats and business logic
	timePattern := regexp.MustCompile(`^([01]\d|2[0-3]):([0-5]\d)$`)
	for i, hour := range req.Hours {
		// Validate day_of_week range
		if hour.DayOfWeek < 0 || hour.DayOfWeek > 6 {
			c.JSON(http.StatusBadRequest, models.APIResponse{
				Success: false,
				Message: "Invalid day_of_week (must be 0-6)",
				Error:   stringPtr("invalid_day_of_week"),
			})
			return
		}

		// Skip time validation if closed
		if hour.IsClosed {
			continue
		}

		// Validate required fields when not closed
		if strings.TrimSpace(hour.OpenTime) == "" {
			c.JSON(http.StatusBadRequest, models.APIResponse{
				Success: false,
				Message: "Open time is required when not closed",
				Error:   stringPtr("open_time_required"),
			})
			return
		}

		if strings.TrimSpace(hour.CloseTime) == "" {
			c.JSON(http.StatusBadRequest, models.APIResponse{
				Success: false,
				Message: "Close time is required when not closed",
				Error:   stringPtr("close_time_required"),
			})
			return
		}

		// Validate time format
		if !timePattern.MatchString(hour.OpenTime) {
			c.JSON(http.StatusBadRequest, models.APIResponse{
				Success: false,
				Message: "Invalid open time format (use HH:MM)",
				Error:   stringPtr("invalid_open_time"),
			})
			return
		}

		if !timePattern.MatchString(hour.CloseTime) {
			c.JSON(http.StatusBadRequest, models.APIResponse{
				Success: false,
				Message: "Invalid close time format (use HH:MM)",
				Error:   stringPtr("invalid_close_time"),
			})
			return
		}

		// T017: Validate that 00:00 is not used for non-closed days
		// 00:00 indicates a closed day and should only be allowed when is_closed is true
		if hour.OpenTime == "00:00" || hour.CloseTime == "00:00" {
			c.JSON(http.StatusBadRequest, models.APIResponse{
				Success: false,
				Message: "00:00 is not a valid time for an open day",
				Error:   stringPtr("invalid_zero_time"),
			})
			return
		}

		// Validate open < close (unless closed)
		if hour.OpenTime >= hour.CloseTime {
			c.JSON(http.StatusBadRequest, models.APIResponse{
				Success: false,
				Message: "Opening time must be before closing time",
				Error:   stringPtr("invalid_time_range"),
			})
			return
		}

		// Store index for better error messages
		_ = i
	}

	// Get restaurant info ID
	var infoID uuid.UUID
	err := h.db.QueryRow("SELECT id FROM restaurant_info LIMIT 1").Scan(&infoID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Message: "Failed to retrieve restaurant info",
			Error:   stringPtr(err.Error()),
		})
		return
	}

	// Update each day's operating hours
	updateQuery := `
		UPDATE operating_hours SET
			open_time = $1, close_time = $2, is_closed = $3, updated_at = NOW()
		WHERE restaurant_info_id = $4 AND day_of_week = $5
	`

	for _, hour := range req.Hours {
		// Normalize time format (handle both HH:MM and ISO datetime)
		openTime := normalizeTimeString(hour.OpenTime)
		closeTime := normalizeTimeString(hour.CloseTime)

		// For closed days, use default times (won't be shown in UI anyway)
		if hour.IsClosed {
			openTime = "00:00"
			closeTime = "00:00"
		}

		_, err := h.db.Exec(
			updateQuery,
			openTime,
			closeTime,
			hour.IsClosed,
			infoID,
			hour.DayOfWeek,
		)
		if err != nil {
			c.JSON(http.StatusInternalServerError, models.APIResponse{
				Success: false,
				Message: "Failed to update operating hours",
				Error:   stringPtr(err.Error()),
			})
			return
		}
	}

	c.JSON(http.StatusOK, models.APIResponse{
		Success: true,
		Message: "Operating hours updated successfully",
		Data: gin.H{
			"updated_count": len(req.Hours),
		},
	})
}

// Helper function to validate phone format (Indonesian format)
func isValidPhone(phone string) bool {
	// Accept formats: +62xxx, 08xxx, 62xxx, or any standard phone format
	phoneRegex := regexp.MustCompile(`^(\+62|62|0)[0-9]{8,13}$`)
	return phoneRegex.MatchString(strings.ReplaceAll(phone, "-", ""))
}

// normalizeTimeString converts various time formats to HH:MM
// Handles: HH:MM, ISO datetime (0000-01-01T10:00:00Z), etc.
func normalizeTimeString(timeStr string) string {
	timeStr = strings.TrimSpace(timeStr)

	// If already in HH:MM format, return as-is
	if matched, _ := regexp.MatchString(`^([01]\d|2[0-3]):([0-5]\d)$`, timeStr); matched {
		return timeStr
	}

	// Try parsing as ISO datetime and extract time
	formats := []string{
		time.RFC3339,          // 0000-01-01T10:00:00Z
		"2006-01-02T15:04:05", // Without timezone
		"15:04:05",            // HH:MM:SS
	}

	for _, format := range formats {
		if t, err := time.Parse(format, timeStr); err == nil {
			return t.Format("15:04") // Return as HH:MM
		}
	}

	// If parsing fails, return original (will fail validation)
	return timeStr
}

File: backend/src/handlers/restaurant-info.ts
import type { Context } from 'hono';
import { sql } from 'drizzle-orm';
import { db } from '../db/connection.js';
import { successResponse, errorResponse } from '../lib/response.js';

// ── UpdateRestaurantInfo ──────────────────────────────────────────────────────

export async function updateRestaurantInfo(c: Context) {
  let body: {
    name: string;
    tagline?: string | null;
    description?: string | null;
    address?: string | null;
    city?: string | null;
    postal_code?: string | null;
    country?: string | null;
    phone?: string | null;
    email?: string | null;
    whatsapp?: string | null;
    map_latitude?: number | null;
    map_longitude?: number | null;
    google_maps_url?: string | null;
    instagram_url?: string | null;
    facebook_url?: string | null;
    twitter_url?: string | null;
    logo_url?: string | null;
    hero_image_url?: string | null;
  };

  try {
    body = await c.req.json();
  } catch {
    return errorResponse(c, 'Invalid request format', 'invalid_json', 400);
  }

  // Validate required fields
  const name = (body.name || '').trim();
  if (!name) {
    return errorResponse(c, 'Restaurant name is required', 'name_required', 400);
  }

  // Validate email format if provided
  if (body.email && body.email.trim()) {
    const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
    if (!emailRegex.test(body.email.trim())) {
      return errorResponse(c, 'Invalid email format', 'invalid_email', 400);
    }
  }

  // Validate phone format if provided
  if (body.phone && body.phone.trim()) {
    const phoneRegex = /^(\+62|62|0)[0-9]{8,13}$/;
    const cleanPhone = body.phone.trim().replace(/-/g, '');
    if (!phoneRegex.test(cleanPhone)) {
      return errorResponse(c, 'Invalid phone format', 'invalid_phone', 400);
    }
  }

  try {
    // Get restaurant info ID (singleton table)
    const infoRes = await db.execute<{ id: string }>(sql`
      SELECT id FROM restaurant_info LIMIT 1
    `);

    if (infoRes.rows.length === 0) {
      return errorResponse(c, 'Failed to retrieve restaurant info', 'not_found');
    }

    const infoId = infoRes.rows[0].id;

    // Update restaurant info
    await db.execute(sql`
      UPDATE restaurant_info SET
        name = ${name},
        tagline = ${body.tagline ?? null},
        description = ${body.description ?? null},
        address = ${body.address ?? null},
        city = ${body.city ?? null},
        postal_code = ${body.postal_code ?? null},
        country = ${body.country ?? null},
        phone = ${body.phone ?? null},
        email = ${body.email ?? null},
        whatsapp = ${body.whatsapp ?? null},
        map_latitude = ${body.map_latitude ?? null},
        map_longitude = ${body.map_longitude ?? null},
        google_maps_url = ${body.google_maps_url ?? null},
        instagram_url = ${body.instagram_url ?? null},
        facebook_url = ${body.facebook_url ?? null},
        twitter_url = ${body.twitter_url ?? null},
        logo_url = ${body.logo_url ?? null},
        hero_image_url = ${body.hero_image_url ?? null},
        updated_at = NOW()
      WHERE id = ${infoId}
    `);

    return successResponse(c, 'Restaurant information updated successfully', { id: infoId });
  } catch (err) {
    return errorResponse(c, 'Failed to update restaurant information', (err as Error).message);
  }
}

// ── UpdateOperatingHours ──────────────────────────────────────────────────────

export async function updateOperatingHours(c: Context) {
  let body: {
    hours: Array<{
      day_of_week: number;
      open_time: string;
      close_time: string;
      is_closed: boolean;
    }>;
  };

  try {
    body = await c.req.json();
  } catch {
    return errorResponse(c, 'Invalid request format', 'invalid_json', 400);
  }

  if (!body.hours || body.hours.length !== 7) {
    return errorResponse(c, 'Operating hours must include all 7 days of the week', 'invalid_hours_count', 400);
  }

  // Validate time formats and business logic
  const timePattern = /^([01]\d|2[0-3]):([0-5]\d)$/;

  for (const hour of body.hours) {
    if (hour.day_of_week < 0 || hour.day_of_week > 6) {
      return errorResponse(c, 'Invalid day_of_week (must be 0-6)', 'invalid_day_of_week', 400);
    }

    if (hour.is_closed) continue;

    if (!(hour.open_time || '').trim()) {
      return errorResponse(c, 'Open time is required when not closed', 'open_time_required', 400);
    }
    if (!(hour.close_time || '').trim()) {
      return errorResponse(c, 'Close time is required when not closed', 'close_time_required', 400);
    }

    if (!timePattern.test(hour.open_time)) {
      return errorResponse(c, 'Invalid open time format (use HH:MM)', 'invalid_open_time', 400);
    }
    if (!timePattern.test(hour.close_time)) {
      return errorResponse(c, 'Invalid close time format (use HH:MM)', 'invalid_close_time', 400);
    }

    if (hour.open_time === '00:00' || hour.close_time === '00:00') {
      return errorResponse(c, '00:00 is not a valid time for an open day', 'invalid_zero_time', 400);
    }

    if (hour.open_time >= hour.close_time) {
      return errorResponse(c, 'Opening time must be before closing time', 'invalid_time_range', 400);
    }
  }

  try {
    // Get restaurant info ID
    const infoRes = await db.execute<{ id: string }>(sql`
      SELECT id FROM restaurant_info LIMIT 1
    `);

    if (infoRes.rows.length === 0) {
      return errorResponse(c, 'Failed to retrieve restaurant info', 'not_found');
    }

    const infoId = infoRes.rows[0].id;

    // Update each day's operating hours
    for (const hour of body.hours) {
      const openTime = hour.is_closed ? '00:00' : normalizeTimeString(hour.open_time);
      const closeTime = hour.is_closed ? '00:00' : normalizeTimeString(hour.close_time);

      await db.execute(sql`
        UPDATE operating_hours SET
          open_time = ${openTime},
          close_time = ${closeTime},
          is_closed = ${hour.is_closed},
          updated_at = NOW()
        WHERE restaurant_info_id = ${infoId} AND day_of_week = ${hour.day_of_week}
      `);
    }

    return successResponse(c, 'Operating hours updated successfully', {
      updated_count: body.hours.length,
    });
  } catch (err) {
    return errorResponse(c, 'Failed to update operating hours', (err as Error).message);
  }
}

// ── Helper: normalizeTimeString ──────────────────────────────────────────────

function normalizeTimeString(timeStr: string): string {
  const trimmed = timeStr.trim();

  // If already in HH:MM format, return as-is
  if (/^([01]\d|2[0-3]):([0-5]\d)$/.test(trimmed)) {
    return trimmed;
  }

  // Try parsing as ISO datetime formats
  const formats = [
    /^\d{4}-\d{2}-\d{2}T(\d{2}:\d{2}):\d{2}/, // ISO 8601
    /^(\d{2}:\d{2}):\d{2}$/,                     // HH:MM:SS
  ];

  for (const fmt of formats) {
    const match = trimmed.match(fmt);
    if (match) return match[1];
  }

  // If parsing fails, return original
  return trimmed;
}
